// play-dl/dist/YouTube/classes/Video.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.YouTubeVideo = void 0;
const Channel_1 = require("./Channel");
/**
 * Class for YouTube Video url
 */
class YouTubeVideo {
    constructor(data) {
        if (!data)
            throw new Error(`Can not initiate ${this.constructor.name} without data`);
        this.id = data.id || undefined;
        this.url = `https://www.youtube.com/watch?v=${this.id}`;
        this.type = 'video';
        this.title = data.title || undefined;
        this.description = data.description || undefined;
        this.durationRaw = data.duration_raw || '0:00';
        this.durationInSec = (data.duration < 0 ? 0 : data.duration) || 0;
        this.uploadedAt = data.uploadedAt || undefined;
        this.views = parseInt(data.views) || 0;
        this.thumbnail = data.thumbnail || {};
        this.channel = new Channel_1.YouTubeChannel(data.channel) || {};
        this.likes = data.ratings?.likes || 0;
        this.dislikes = data.ratings?.dislikes || 0;
        this.live = !!data.live;
        this.private = !!data.private;
        this.tags = data.tags || [];
    }
    get toString() {
        return this.url || '';
    }
    get toJSON() {
        return {
            id: this.id,
            url: this.url,
            title: this.title,
            description: this.description,
            durationInSec: this.durationInSec,
            durationRaw: this.durationRaw,
            uploadedAt: this.uploadedAt,
            thumbnail: this.thumbnail,
            channel: this.channel,
            views: this.views,
            type: this.type,
            tags: this.tags,
            ratings: {
                likes: this.likes,
                dislikes: this.dislikes
            },
            live: this.live,
            private: this.private
        };
    }
}
exports.YouTubeVideo = YouTubeVideo;
